cc3d/tests/pycorespecs_tester.py

```diff
@@ -0,0 +1,1565 @@
+
+    # Validation error: undefined cell type
+    volume_specs.param_new("T3", target_volume=1.0, lambda_volume=1.0)
+    try:
+        volume_specs.validate(*specs)
+        return name, False
+    except SpecValueError as e:
+        print(name, str(e), str(e.names))
+        cell_type_specs.cell_type_append("T3")
+        try:
+            volume_specs.validate(*specs)
+        except SpecValueError as e:
+            print(name, str(e), str(e.names))
+            return name, False
+
+    return name, True
+
+
+def test_validation():
+    """
+    Tests validation
+
+    :return: None
+    """
+    print("Testing validation")
+    validation_tests = [
+        test_validation_adhesion_flex,
+        test_validation_blob_initializer,
+        test_validation_chemotaxis,
+        test_validation_connectivity_global,
+        test_validation_contact,
+        test_validation_contact_internal,
+        test_validation_contact_local_flex,
+        test_validation_curvature,
+        test_validation_diffusion_solver_fe,
+        test_validation_external_potential,
+        test_validation_focal_point_plasticity,
+        test_validation_kernel_diffusion_solver,
+        test_validation_length_constraint,
+        test_validation_pde_boundary_condition,
+        test_validation_potts,
+        test_validation_reaction_diffusion_solver_fe,
+        test_validation_secretion,
+        test_validation_steady_state_diffusion_solver,
+        test_validation_surface,
+        test_validation_uniform_initializer,
+        test_validation_volume
+    ]
+
+    with _SpecValueErrorContextBlock() as err_ctx:
+        for vt in validation_tests:
+            with err_ctx.ctx:
+                name, result = vt()
+                if result:
+                    print("Validation passed:", name)
+                else:
+                    raise SpecValueError("Validation failed: " + name)
+
+
+def test_serial() -> None:
+    """
+    Tests whether all main specs can be serialized
+
+    :return: None
+    """
+    from cc3d.core.PyCoreSpecs import PLUGINS, STEPPABLES, INITIALIZERS, Metadata
+    from cc3d.core.PyCoreSpecs import PIFDumperSteppable, PIFInitializer
+    import pickle
+
+    ins = {PIFDumperSteppable: ([], {"pif_name": "dummy", "frequency": 1}),
+           PIFInitializer: ([], {"pif_name": "dummy"})}
+
+    for spec in PLUGINS + STEPPABLES + INITIALIZERS + [Metadata]:
+        print("Testing serialization:", spec)
+
+        try:
+            args, kwargs = ins[spec]
+        except KeyError:
+            args, kwargs = [], {}
+
+        o = spec(*args, **kwargs)
+        dumpo = pickle.dumps(o)
+        loado = pickle.loads(dumpo)
+        print(o, loado, dumpo)
+
+
+def main():
+    """
+    Performs tests
+
+    :return: None
+    """
+    with _SpecValueErrorContextBlock() as err_ctx:
+        err_ctx.ctx(test_import)
+        err_ctx.ctx(test_validation)
+        err_ctx.ctx(test_serial)
+
+
+if __name__ == "__main__":
```

Comment:
I am thinking that perhaps in future iterations would should move to pytest framework.